evap/student/forms.py

```diff
@@ -22,23 +59,22 @@ def __init__(self, *args, contribution, questionnaire, **kwargs):
         self.questionnaire = questionnaire
 
         for question in self.questionnaire.questions.all():
-            # generic arguments for all kinds of fields
-            field_args = dict(label=question.text)
-
             if question.is_text_question:
-                field = forms.CharField(required=False, widget=forms.Textarea(),
-                                        **field_args)
+                field = TextAnswerField(label=question.text)
             elif question.is_rating_question:
-                choices = CHOICES[question.type]
-                field = forms.TypedChoiceField(widget=forms.RadioSelect(attrs={'choices': choices}),
-                                               choices=zip(choices.values, choices.names),
-                                               coerce=int,
-                                               **field_args)
+                field = RatingAnswerField(
+                    widget_choices=CHOICES[question.type],
+                    choices=zip(CHOICES[question.type].values, CHOICES[question.type].names),
+                    label=question.text,
+                    allows_textanswer=question.allows_textanswers,
```

Comment:
if you made RatingAnswerField (and perhaps TextAnswerField) only take a question as a parameter, the code would probably get simpler. They would also be less generic, but then again they are not meant to be generic.